### PR TITLE
Add typechecking and declaration output

### PR DESCRIPTION
**Pull request by sirreal**
Add type checking and typescript declaration file generation to the plugin.

We throw some errors in places where types suggest we may have problems. Especially `publicPath`, we only allow a string but webpack options permit several variations.

### CODE REVIEW

File: src/plugin.js [hunk 1]
```diff
@@ -111,32 +122,49 @@ const defaultGetIconSize = (fileName) => {
...
+ * @property {unknown} [name]
+ * @property {unknown} [short_name]
+ * @property {unknown} [start_url]
+ * @property {unknown} [display]
+ * @property {unknown} [background_color]
+ * @property {unknown} [theme_color]
+ * @property {unknown} [description]
+ * @property {unknown} [icons]
+ * @property {unknown} [related_applications]
+ * @property {unknown} [prefer_related_applications]
```
Comment (sirreal) [inline src/plugin.js]: Tempting to parse to this format:

https://developer.mozilla.org/en-US/docs/Web/Manifest

File: package.json [hunk 2]
```diff
@@ -3,8 +3,12 @@
   "version": "3.0.2",
   "description": "webpack-web-app-manifest-plugin is a webpack plugin that generates a PWA manifest and integrates with the assets JSON generated by assets-webpack-plugin.",
   "main": "index.js",
+  "types": "types",
   "scripts": {
-    "test": "jest",
+    "build": "yarn tsc",
```
Comment (blowery) [inline package.json:7]: do you need the `yarn` here?

File: src/plugin.js (hunk 1 above)
Comment (blowery) [reply src/plugin.js]: https://www.npmjs.com/package/@types/web-app-manifest

**Comment by blowery** [comment]
Can / should we just flip the plugin to a `.ts` file?

File: package.json (hunk 2 above)
Comment (sirreal) [reply package.json:7]: Not strictly necessary, no. [However, Yarn 2 encourages this](https://yarnpkg.com/getting-started/migration#call-binaries-using-yarn-run-rather-than-node_modulesbin).

**Comment by sirreal** [comment]
It is a bit tricky to actually switch to TypeScript sources. There's some mechanical translation, but also some infrastructure things to fix up. I had to add some babel configuration and dev dependencies so that Jest's transformers will transform source for testing. I've done that now.

I also pulled in https://www.npmjs.com/package/@types/web-app-manifest and I dropped the validation function that was removing unknown properties and `undefined` values. Instead, we can rely on the type system.

**Comment by sirreal** [comment]
I went ahead and landed this. It's been working well in my testing. I'm happy to circle back and address any feedback.

**Pull request merged**